internal/internal_task_handlers.go

```diff
@@ -1819,6 +1852,33 @@ func (ath *activityTaskHandlerImpl) Execute(taskList string, t *s.PollForActivit
 
 	ctx, span := createOpenTracingActivitySpan(ctx, ath.tracer, time.Now(), activityType, t.WorkflowExecution.GetWorkflowId(), t.WorkflowExecution.GetRunId())
 	defer span.Finish()
+
+	if activityImplementation.GetOptions().EnableAutoHeartbeat && t.HeartbeatTimeoutSeconds != nil && *t.HeartbeatTimeoutSeconds > 0 {
+		go func() {
+			autoHbInterval := time.Duration(*t.HeartbeatTimeoutSeconds) * time.Second / 2
+			ticker := time.NewTicker(autoHbInterval)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ath.workerStopCh:
+					return
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
+					hbErr := invoker.BackgroundHeartbeat()
+					if hbErr != nil && IsCanceledError(hbErr) {
+						ath.logger.Error("Activity auto heartbeat error.",
```

Comment:
nit: shall we also record errors other than canceled?

Reply:
Yes. I actually only want to log error if it's not cancelledError. Updated.